Build locally msi (#6067)
jenkins-infra/jenkins.io: content/blog/2023/02/2023-02-27-build-msi-locally.adoc

=== Git

There are quite a few ways to get Git on Windows, but the most straightforward is to see what the https://git-scm.com/download/win[official Git website recommends].

=== Install MSBuild

You can install https://aka.ms/vs/17/release/vs_BuildTools.exe[MSBuild] from Visual Studio or from the https://visualstudio.microsoft.com/downloads/#build-tools-for-visual-studio-2022[Build Tools for Visual Studio].

This command line tool is used to build the MSI file.

=== Install .NET Framework 3.5

You may already have it installed on your machine, but not activated.
You can activate it from the Windows Features dialog box.

To access this dialog box, press the keys +++<kbd>+++⊞ Win+++</kbd>+++ + +++<kbd>+++R+++</kbd>+++, then enter the command `appwiz.cpl` and push enter.
Search for

____
Turn Windows features on or off.
____

Tick the `.NET Framework 3.5` entry and install.

IMPORTANT: now run Windows Update to check for security updates.

If it is not installed yet, you can install https://dotnet.microsoft.com/en-us/download/dotnet-framework/net35-sp1[.NET Framework 3.5] from the Windows Features.

=== Check if you have PowerShell, install it if not

> In recent versions of Windows, PowerShell is already installed and accessible through the https://support.microsoft.com/en-us/topic/6453ce98-da91-476f-8651-5c14d5777c20#:~:text=In%20Windows%2011%2022H2%2C%20the,an%20instance%20of%20Windows%20Terminal[terminal] application.
> At the time of writing, the pre-installed version is `5.1.22621.963`.
> You can also install the latest version from the https://www.microsoft.com/en-us/p/powershell/9mz1snwt0n5d?activetab=pivot:overviewtab[Microsoft Store] (7.3.2 at the time of writing). +
> You could also install PowerShell from GitHub by issuing the following command:

Reviewer: ```suggestion
In recent versions of Windows, PowerShell is already installed and accessible through the link:https://support.microsoft.com/en-us/windows/command-prompt-and-windows-powershell-for-windows-11-6453ce98-da91-476f-8651-5c14d5777c20[terminal] application.
At the time of writing, the pre-installed version is `5.1.22621.963`.
You can also install the latest version from the link:https://apps.microsoft.com/store/detail/powershell/9MZ1SNWT0N5D[Microsoft Store] (7.3.2 at the time of writing). +

You could also install PowerShell from GitHub by issuing the following command:
```
small suggestion to update the links to macros, if the links are incorrect please ignore

[,powershell]
----
winget show "Microsoft.PowerShell" -s winget
----

> This would give out an output similar to:

Reviewer: ```suggestion
This would give an output similar to:
```
small suggestion for syntax

Moniker: pwsh
Description:
  PowerShell is a cross-platform (Windows, Linux, and macOS) automation and configuration tool/framework that works well with your existing tools and is optimized for dealing with structured data (e.g. JSON, CSV, XML, etc.), REST APIs, and object models.
  It includes a command-line shell, an associated scripting language and a framework for processing cmdlets.
Homepage: https://microsoft.com/PowerShell
License: MIT
License Url: https://github.com/PowerShell/PowerShell/blob/master/LICENSE.txt
Copyright: Copyright (c) Microsoft Corporation
Copyright Url: https://github.com/PowerShell/PowerShell/blob/master/LICENSE.txt
Release Notes Url: https://github.com/PowerShell/PowerShell/releases/tag/v7.3.2
Tags:
  command-line
  cross-platform
  open-source
  powershell
  pwsh
  shell
Installer:
  Installer Type: wix
  Installer Url: https://github.com/PowerShell/PowerShell/releases/download/v7.3.2/PowerShell-7.3.2-win-x64.msi
  Installer SHA256: a4f7d081c5f74bc8d6c75f1dfee382b7fd9335361181748fee590ecdbc96cb26
  Release Date: 2023-01-24
----

You can see that the latest version is `7.3.2` and that the installer is a `.msi` file located on GitHub.

> Just follow this link with your browser, and install PowerShell from this file once downloaded.

Reviewer: ```suggestion
Just follow the link provided with your browser and install PowerShell from this file once it has downloaded.
```
small suggestion for syntax

== Build the MSI

=== Clone the Jenkins packaging repository

Choose your `git` tool and clone the https://github.com/jenkinsci/packaging.git[Jenkins packaging repository] on your machine.

=== Prepare the build

> Open a `terminal` and go to the folder where you cloned the repository. For example `C:\jenkinsci\packaging\`.
> You now have to declare where you downloaded the Jenkins war file so the build script can find it.

Reviewer: ```suggestion
Open a `terminal` window and go to the folder where you cloned the repository.
For example `C:\jenkinsci\packaging\`.
You now have to declare where you downloaded the Jenkins war file, so the build script can find it.
```
small suggestion for syntax/formatting

[,powershell]
----
$env:War = "$env:USERPROFILE\jenkins.war"
----

> If you ever moved it into your repository clone folder, you can use this command instead:

Reviewer: ```suggestion
If you have previously moved it into your repository clone folder, you can use this command instead:
```
suggestion for syntax, doesn't need to be applied if this is incorrect

  Windows Installer XML Toolset Linker version 3.11.1.2318
  Copyright (c) .NET Foundation and contributors. All rights reserved.

  jenkins -> C:\jenkinsci\packaging\msi\build\bin\Release\en-US\jenkins-2.392.msi
Done Building Project "C:\jenkinsci\packaging\msi\build\jenkins.wixproj" (default targets).

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:08.26
----

== Locate the generated MSI file

The MSI file is located in the `.\bin\Release\en-US\` folder.

> You will find there the generated MSI file and its `sha256` file.

Reviewer: ```suggestion
In this folder, you will find the generated MSI file and its `sha256` file.
```
small suggestion for syntax/formatting

---
layout: post
title: "How to build an unsigned Jenkins MSI on your Windows machine"
tags:
- jenkins
- windows
- msi
authors:

> - gounthar

Reviewer 2: ```suggestion
- gounthar
- notmyfault
```
> Don't hesitate to add your handles to the authors' part folks. 🤷

For sure :)

discourse: true
---

Should you ever need to rebuild a Jenkins MSI on your Windows machine, here is a way to do it.

== Pre-requisites

=== Jenkins WAR file

First of all, you should get the Jenkins war file that will be inside that MSI file.
You can get it from the official Jenkins website or from https://updates.jenkins.io/[the Jenkins update center].

Check https://www.jenkins.io/download/[Jenkins download page] and download the latest weekly version of Jenkins for example.
You could always hit https://updates.jenkins.io/latest/jenkins.war[the direct link] to get the latest weekly version, but then you wouldn't know which version number you are using. Just saying.

=== Git

There are quite a few ways to get Git on Windows, but the most straightforward is to see what the https://git-scm.com/download/win[official Git website recommends].
You have a few other options here:

- Install https://community.chocolatey.org/packages/git[Git for Windows]
- Use your IDE to clone the repository
- Use https://desktop.github.com/[GitHub Desktop]
- Use https://www.sourcetreeapp.com/[SourceTree]

> - Use Cygwin integrated https://cygwin.com/packages/summary/git.html[Git command]

Reviewer 2: Similar to the packaging docs, I think listing one way to obtain git is enough.

Author: At least I tried. 🤷 🤣
